### ISSUE TITLE
Problem loading JDBC driver while debugging Karate test.

### ISSUE THREAD

**Issue by djangofan** [kirksl/karate-runner#53]
Problem loading JDBC driver while debugging Karate test.

Instructions and steps to reproduce are described in this ticket:
https://github.com/intuit/karate/issues/1162

I simplified the instructions and the [example project I supplied](https://github.com/djangofan/karate-db-sample) to make it as easy as possible to reproduce.

**Comment by kirksl** [comment]
@djangofan happy to take a look at this if you can help me pull down your example project repro'ing the issue.  when clicking the link above i get -->
![image](https://user-images.githubusercontent.com/24948266/83780086-8f91bf00-a641-11ea-98b9-cdefa5c608f9.png)

**Comment by djangofan** [comment]
Fixed.  I had to change repo to be public because Github changed repos to default to private. :-(

**Comment by kirksl** [comment]
no worries.  just pulled it down and ran the steps below.  seems to work?

1. ran mvn clean test and it works
2. opened mssql.feature and set breakpoint on line 10
3. clicked karate: debug codelens and it failed (see karate.debug.error image)
4. deleted launch.json under .vscode folder as launch config looks bad
5. created new launch.json using default karate (debug): maven config
6. clicked karate: debug codelens again and it succeeded (see karate.debug.no.error image)

![karate debug error](https://user-images.githubusercontent.com/24948266/83788312-dbe0fd00-a649-11ea-93ad-134d260a3743.png)
![karate debug no error](https://user-images.githubusercontent.com/24948266/83788317-dbe0fd00-a649-11ea-8a7b-4baa2f08d79a.png)

**Comment by djangofan** [comment]
Thanks @kirksl  , I see what you mean.

So, I have multiple test suites, and I am in the habit of saving multiple run configs in launch.json that I check-in to my projects so that other people can use pre-configured run configurations.

Here is an example of How I tried changing the run configuration, based on your suggestion above, to regenerate:

```
OLD Run Config :
mvn test-compile -f pom.xml exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args='-d' -Dexec.classpathScope=test -Dtest.env=int -DDB.password=passWrd -Dtest.type=reg

NEW Run Config:
${config:karateRunner.karateJar.commandLineArgs} -d -Dtest.env=int -DDB.password=passWrd -Dtest.type=reg
```

So, how does the new Run config know about the Maven classpath?  Do I need to first create a .jar of my project?

**Comment by djangofan** [comment]
Just FYI.  When I create a new launch config, using latest VSCode Karate-Runner, I do not see the option you describe in this step:

5. created new launch.json using default karate (debug): maven config

Can you clarify?

**Comment by djangofan** [comment]
Ok, after I had established a `launch.json`, when I choose "add configuration" ONLY in that case do I see the option to add a "maven debug config".  Just FYI, if you want to fix that.

So, the auto-generated run config is:

`mvn test-compile -f \"${command:karateRunner.getDebugBuildFile}\" exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args=\"-d\" -Dexec.classpathScope=\"test\" ${config:karateRunner.karateRunner.commandLineArgs}`

![Screen Shot 2020-06-04 at 1 30 56 PM](https://user-images.githubusercontent.com/810894/83807371-d692ab00-a667-11ea-8c30-88352e195170.png)

**Comment by djangofan** [comment]
I uploaded a video that shows you what it looks like to me:
https://youtu.be/xsxoz4s7TpM

**Comment by kirksl** [comment]
@djangofan let me know if this (image from the karate runner readme on the marketplace) solves your troubles.  sounds like it just wasn't clear how to setup your launch config?
![image](https://user-images.githubusercontent.com/24948266/83823336-a27bb200-a688-11ea-8fe6-feba392f898c.png)

**Comment by djangofan** [comment]
I just read through that.  There is nothing there different that what I described as my steps to reproduce.   I've been using Karate every day since January.  No confusion there.

**Comment by kirksl** [comment]
the problem is that you forgot to "pick" the maven debug configuration as shown in the red boxed step from the instructions.  what you did was just merely add it to the launch.json.  once added to the json you have to pick between all of the configs within that file via the dropdown that i've red boxed too.  let me know if that solves your problem.
![image](https://user-images.githubusercontent.com/24948266/83827102-52095200-a692-11ea-9451-b408fffd1a4d.png)

**Comment by kirksl** [comment]
@djangofan looks like you forgot to pick the maven debug config from the vscode dropdown.  you added to the launch.json but then you are left with multiple debug configs and left it set to the standalone config instead of the maven config.
![image](https://user-images.githubusercontent.com/24948266/83827255-b3312580-a692-11ea-911e-1dda59d3d78b.png)

**Comment by djangofan** [comment]
Yes, good catch.  I didn't notice I did that in the heat of recording the video.    Here is what happens between those 2 Run Configurations:

1. The Karate (debug): Standalone -  hangs and finally gives up after about 30 seconds
2. The Karate (debug): Maven -  runs and at the end, in the "Debug Console" you can see the class loading error.

**Comment by kirksl** [comment]
@djangofan for 1. that's expected.  standlone debug mode is only when you are using karate.jar not maven.  since it never finds karate.jar it times out.  for 2. i'm not repro'ing a class loading error.  see image.  if you're still repro'ing it please give me exact repro steps.  at this point this really doesn't seem like an issue with karate runner as it's merely just assembling commands on the command line so that end users don't have to type everything.
![image](https://user-images.githubusercontent.com/24948266/83959883-05ab4700-a837-11ea-9c36-e5961820a9bc.png)

**Comment by djangofan** [comment]
kirksl:

2-3 weeks ago your runner worked with this same example code  but does not
work today.  i don’t know the exact day it stipped working.   do you have a
way that I can install the older plugin?   if so,  then i can verify the
regression.

ill record another video tomorrow that is much more detailed

-jon

**Comment by kirksl** [comment]
yes you can install any previous version
![image](https://user-images.githubusercontent.com/24948266/83961733-cd156880-a84a-11ea-9ded-cae0bb212e62.png)

**Comment by djangofan** [comment]
Thank you.   Here is what I found today:

1.   Changing to the older Karate-Runner 1.8.6 didn't resolve the problem.   So, I went back to latest version.   Therefore, whatever this is, it's very possible it is not a regression.

2.  I tried using Gradle and did not get a different result.

3.  I tried switching to  JUnit 4 instead of Junit5 and it also didn't make a difference.

4.  I checked out the master branch of karate.git  and I altered the Dogs.feature example to load the MSSQL example instead and I AM ABLE TO REPRODUCE with that.

Sorry, the above investigations took me a while  and I didn't record another video.     My issue is still unresolved though.    I am glad I can now reproduce using the Karate DEMO.

To reproduce using Karate DEMO, reduce the Dogs.feature file to this:

```
Scenario: create and retrieve a dog

# use jdbc to validate
* def config = { username: 'sa', password: '', url: 'jdbc:sqlserver://localhost:1433;databaseName=TEST', driverClassName: 'com.microsoft.sqlserver.jdbc.SQLServerDriver' }
* def DbUtils = Java.type('com.intuit.karate.demo.util.DbUtils')
* def db = new DbUtils(config)

# since the DbUtils returns a Java Map, it becomes normal JSON here !
# which means that you can use the full power of Karate's 'match' syntax
* def dogs = db.readRows('SELECT TOP (2) * FROM [TEST].[dbo].[Order]')
* match dogs contains { ID: '#(id)', NAME: 'Scooby' }
```

Then use this in you pom.xml:

```
        <dependency>
            <groupId>com.microsoft.sqlserver</groupId>
            <artifactId>mssql-jdbc</artifactId>
            <version>8.2.2.jre8</version>
            <scope>test</scope> 
        </dependency>  
```

**Comment by djangofan** [comment]
@kirksl  @ptrthomas   I finally figured it out.    The fix is related to:
http://www.kfu.com/~nsayer/Java/dyn-jdbc.html

Here is the exact change that fixed it for me:
https://github.com/djangofan/karate-db-sample/blob/e9e3a5d6fad8a5be9f75799405b48966f2552a49/src/test/java/qa/test/DbUtils.java#L36


SO, basically I believe Karate-Runner is perhaps forking away from the System Classloader and therefore the JDBC class will not load from there?   Can you confirm?

**Comment by ptrthomas** [comment]
<img width="503" alt="image" src="https://user-images.githubusercontent.com/915480/84220228-eee34980-aaef-11ea-8904-134e8997c647.png">

we start a new `Thread` - I don't think this code has changed ever since we started. feel free to investigate and suggest changes: https://github.com/intuit/karate/blob/develop/karate-core/src/main/java/com/intuit/karate/debug/DapServerHandler.java#L345

**Comment by djangofan** [comment]
@ptrthomas   I am not completely sure.   My best guess is that in your Runnable,  which you named com.intuit.karate.Runner,  you pass `Thread.currentThread().getContextClassLoader()` at line 304    rather than `.getSystemClassLoader()` which basically results in the JDBC DriverManger denying the scan to load the driver??    I don't have a lot of the context to resolve this myself.    I bet if you had a way to temporarily skip the call to .parallel, and instead run serially,  then the driver would have loaded because then you wouldn't be using the contextClassLoader?

That is my best guess.   If you don't fix it, then I would document this for people so they know the workaround.    If you do decide to fix it, as I said earlier, I was able to reproduce the problem by hacking your Dogs.feature to load MSSQL.

**Comment by ptrthomas** [comment]
> I don't have a lot of the context to resolve this myself

@djangofan I think you do. just use the developer guide and it will take you 5 minutes to confirm what change we need to make: https://github.com/intuit/karate/wiki/Developer-Guide

**Comment by djangofan** [comment]
@kirksl   yes, i believe its more related to Karate .parallel execution.  I need to take some time to reproduce and ill make a new story on karate project when i am ready.

**Comment by djangofan** [comment]
@ptrthomas I updated my example project so that it now has 2 karate tests in a `mssql.feature` file that reproduce both the bug and the workaround:  https://github.com/djangofan/karate-db-sample/blob/master/src/test/java/features/mssql.feature

My next step is to get your latest Karate code (develop branch), and wire up this above example on that code and then try to debug karate code locally to figure out where it is breaking.    I'll do that hopefully this weekend.

**Comment by lhimo** [comment]
@djangofan I get the same error, did you manage to fix it?
